[libc++][ranges] Fix `ranges::join_view` segmented iterator trait

lbonn opened the pull request:
The outer iterator needs to move to the next segment when calling __compose.

Without this change, `find_segment_if` would never reach the end of the join_view which caused erroneous result when calling `ranges::find` on a join_view of bidirectional ranges.

Other specializations using the segmented iterator trait were likely to be affected as well.

Fixes #158279
Fixes #93180

lbonn commented:
Not sure what is going on with the macos failures, I don't think they are touching these changes.

```
   Failed Tests (9):
    llvm-libc++-shared.cfg.in :: std/re/re.alg/re.alg.match/awk.locale.pass.cpp
    llvm-libc++-shared.cfg.in :: std/re/re.alg/re.alg.match/basic.locale.pass.cpp
    llvm-libc++-shared.cfg.in :: std/re/re.alg/re.alg.match/ecma.locale.pass.cpp
    llvm-libc++-shared.cfg.in :: std/re/re.alg/re.alg.match/extended.locale.pass.cpp
    llvm-libc++-shared.cfg.in :: std/re/re.alg/re.alg.search/awk.locale.pass.cpp
    llvm-libc++-shared.cfg.in :: std/re/re.alg/re.alg.search/basic.locale.pass.cpp
    llvm-libc++-shared.cfg.in :: std/re/re.alg/re.alg.search/ecma.locale.pass.cpp
    llvm-libc++-shared.cfg.in :: std/re/re.alg/re.alg.search/extended.locale.pass.cpp
    llvm-libc++-shared.cfg.in :: std/re/re.traits/lookup_collatename.pass.cpp
```

They already have some XFAIL on some other linux and freebsd targets.

philnik777 commented on libcxx/test/std/ranges/range.adaptors/range.join/range.join.iterator/find.pass.cpp:
```diff
@@ -0,0 +1,37 @@
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17
+
+#include <algorithm>
```
This should be in the algorithm tests, since that's where the optimization is.

philnik777 commented on libcxx/test/std/ranges/range.adaptors/range.join/range.join.iterator/find.pass.cpp:
```diff
@@ -0,0 +1,37 @@
+
+constexpr bool test() {
+  // Test the segmented iterator implementation of join_view
+  // https://github.com/llvm/llvm-project/issues/158279
```
```suggestion
  // https://llvm.org/PR158279
```

philnik777 commented on libcxx/include/__ranges/join_view.h:
```diff
@@ -255,7 +255,18 @@ struct join_view<_View>::__iterator final : public __join_view_iterator_category
     requires forward_range<_Base>
-      : __outer_(std::move(__outer)), __inner_(std::move(__inner)), __parent_(__parent) {}
+      : __outer_(std::move(__outer)), __inner_(std::move(__inner)), __parent_(__parent) {
+    auto __get_inner_range = [&]() -> decltype(auto) {
+      if constexpr (__ref_is_glvalue)
+        return *__get_outer();
+      else
+        return *__parent_->__inner_;
+    };
+    if (__inner_ == ranges::end(std::__as_lvalue(__get_inner_range()))) {
+      ++__get_outer();
+      __satisfy();
+    }
+  }
```
It seems very weird to me to have logic inside a constructor that simply initializes the members. IMO it'd be much clear _why we do this logic if it's inside the segmented iterator traits. They should be exactly as complex as they need to be.

lbonn replied:
Updated

lbonn committed:
[libc++][ranges] Fix `ranges::join_view` segmented iterator trait

The outer iterator needs to move to the next segment when calling
__compose.

Without this change, `find_segment_if` would never reach the end of the
join_view which caused erroneous result when calling `ranges::find` on a
join_view of bidirectional ranges.

philnik777 reviewed (approved):
Thanks! Do you need me to commit this?

frederick-vs-ja reviewed (approved):
LGTM! Merging as CI failures are unrelated.

frederick-vs-ja commented:
@lbonn Could you make you mail address public for GitHub for next contribution(s)? We conventionally expect public mail address on GitHub.

lbonn commented:
@frederick-vs-ja oh no, sorry about that! I've made the change for next time.
